pkg/virt-launcher/virtwrap/storage/backup.go

```diff
@@ -0,0 +1,288 @@
+	if exists && backupMetadata.Name != "" {
+		// Same start time is a unique indication for the backup
+		// since backupname can be reused
+		if backupMetadata.StartTimestamp == backupOptions.BackupStartTime {
+			if backupMetadata.EndTimestamp == nil {
+				// backup is already in progress, ignore
+				return true, nil
+			} else {
+				// backup already completed should not initialize the same backup again
+				return false, fmt.Errorf("backup %s that started at %s already executed, finished at %v, completed: %t",
+					backupOptions.BackupName, *backupMetadata.StartTimestamp, *backupMetadata.EndTimestamp, backupMetadata.Completed)
+			}
+		} else {
+			if backupMetadata.EndTimestamp == nil {
+				// another backup already exists and has not completed yet
+				return false, fmt.Errorf("backup %s already in progress, need to wait for completion", backupMetadata.Name)
+			}
+			// Old backup has completed, allow new backup to overwrite the metadata
+			log.Log.Infof("Previous backup %s completed at %v, initializing new backup %s",
+				backupMetadata.Name, backupMetadata.EndTimestamp, backupOptions.BackupName)
+		}
+	}
+
+	b := api.BackupMetadata{
+		Name:           backupOptions.BackupName,
+		StartTimestamp: backupOptions.BackupStartTime,
+		SkipQuiesce:    backupOptions.SkipQuiesce,
+	}
+	m.metadataCache.Backup.Store(b)
+	log.Log.V(3).Infof("initialize backup metadata: %v", b)
+
+	return false, nil
+}
+
+func (m *StorageManager) backup(vmi *v1.VirtualMachineInstance, backupOptions *backupv1.BackupOptions) (failed error) {
+	logger := log.Log.Object(vmi)
+	domName := api.VMINamespaceKeyFunc(vmi)
+	dom, err := m.virConn.LookupDomainByName(domName)
+	if dom == nil || err != nil {
+		return err
+	}
+	defer dom.Free()
+
+	domainDisks, err := util.GetAllDomainDisks(dom)
+	if err != nil {
+		log.Log.Object(vmi).Reason(err).Error("failed to parse domain XML to get disks.")
+		return err
+	}
+
+	var backupPath string
+	if backupOptions.PushPath != nil {
+		backupPath = getBackupPath(backupOptions, vmi.Name)
+		if err := kutil.MkdirAllWithNosec(backupPath); err != nil {
+			return fmt.Errorf("error creating dir for backup: %v", err)
+		}
+		defer func(path string) {
+			if failed != nil {
+				logger.Reason(failed).Error("failed to run backup")
+				if err := os.RemoveAll(path); err != nil {
+					logger.Reason(err).Error("failed to clean up backup directory")
+				}
+			}
+		}(backupPath)
+	}
+	domainBackup, domainCheckpoint := generateDomainBackup(domainDisks, backupOptions, backupPath)
+	backupXML, err := xml.Marshal(domainBackup)
+	if err != nil {
+		logger.Reason(err).Error("marshalling backup xml failed")
+		return err
+	}
+	checkpointXML, err := xml.Marshal(domainCheckpoint)
+	if err != nil {
+		logger.Reason(err).Error("marshalling checkpoint xml failed")
+		return err
+	}
+
+	frozenFS := false
+	if !backupOptions.SkipQuiesce {
+		logger.Info("Freezing VMI to capture backup state")
+		if err := dom.FSFreeze(nil, 0); err != nil {
```

Comment:
Calling `dom.FSFreeze` here reminded me of #13682 where we added sync() calls for backend PVCs when FSFreeze API is called. This got me thinking about persistent TPM and backup integration. Is there anything we can/should do there?